File: LC/cpp/020_LC_ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        
        vector<char> stack;
        map<char, char> pairs = 
        {{'{', '}'}, {'[', ']'}, {'(', ')'}};

        for(int i = 0; i < s.length(); i++)
        {
            auto it = pairs.find(s[i]);
            if(it != pairs.end()){
                // next character is an opening character
                stack.push_back(s[i]);
            }
            else if(!stack.empty() && pairs[stack[stack.size()-1]] == s[i]){
                // next character matches the last opening character
                stack.pop_back();
            }
            else{
                return false;
            }
        }

        if(stack.empty()){
            return true;
        }
        return false;

    }
};

File: LC/python/020_LC_ValidParentheses.py
#!/usr/bin/env python3

class Solution:
    def isValid(self, s: str) -> bool:

        stack = []
        pairs = { '{':'}', '[':']', '(':')'}
        for c in s:
            if(c in pairs.keys()):
                stack.append(c)
            elif(stack and pairs[stack[len(stack)-1]] == c):
                stack.pop()
            else:
                return False

        if(len(stack) == 0):
            return True
        return False
